core/drivers/dcp/imx_dcp.c

```diff
@@ -0,0 +1,1035 @@
+#include <mm/core_memprot.h>
+#include <tee/cache.h>
+#include <utee_defines.h>
+
+/* Virtual DCP base address */
+static vaddr_t dcp_base;
+/* Driver state: initialized or not */
+static bool driver_init;
+
+/* Spinlock used to lock SRAM key register */
+unsigned int key_store_spinlock = SPINLOCK_UNLOCK;
+uint32_t key_store_except;
+
+/* Spinlock used to lock clock access */
+unsigned int clock_spinlock = SPINLOCK_UNLOCK;
+uint32_t clock_except;
+
+struct lock_channel {
+	struct mutex m; /* Mutex used to lock channels registers */
+};
+
+static struct lock_channel lock_channel[DCP_NB_CHANNELS] = {
+	{
+		/* Channel 0 */
+		.m = MUTEX_INITIALIZER,
+	},
+	{
+		/* Channel 1 */
+		.m = MUTEX_INITIALIZER,
+	},
+	{
+		/* Channel 2 */
+		.m = MUTEX_INITIALIZER,
+	},
+	{
+		/* Channel 3 */
+		.m = MUTEX_INITIALIZER,
+	},
+};
+
+/*
+ * Counter used with clock_spinlock
+ * It avoids:
+ *	- useless clock activation (i.e. when the clock was activated by
+ *	  another channel)
+ *	- to disable clock whereas clock is needed by another channel
+ */
+static unsigned int clock_refcount;
+
+static const struct hashalg hash_alg[3] = {
+	{
+		/* SHA1*/
+		.type = DCP_CONTROL1_HASH_SELECT_SHA1,
+		.size = TEE_SHA1_HASH_SIZE,
+	},
+	{
+		/* SHA256 */
+		.type = DCP_CONTROL1_HASH_SELECT_SHA256,
+		.size = TEE_SHA256_HASH_SIZE,
+	},
+	{
+		/* CRC32 */
+		.type = DCP_CONTROL1_HASH_SELECT_CRC32,
+		.size = DCP_CRC32_SIZE,
+	},
+};
+
+/* Compatible DCP string for i.MX6ULL/ULZ/SL/SLL device tree */
+#ifdef CFG_DT
+static const char *const dt_ctrl_match_table[] = {
+	"fsl,imx28-dcp",
+	"fsl,imx6sl-dcp",
+};
+#endif
+
+/* Memory used for context switching */
+static struct align_buf hw_context_buffer;
+
+/*
+ * Enable/disable the DCP clock
+ *
+ * @enable   Enable the clock if true, disable if false
+ */
+static void dcp_clk_enable(bool enable)
+{
+	uint32_t reg = 0;
+	vaddr_t ccm_base = 0;
+
+	ccm_base = core_mmu_get_va(CCM_BASE, MEM_AREA_IO_SEC);
+	reg = io_read32(ccm_base + CCM_CCGR0);
+
+	if (enable)
+		reg |= DCP_CLK_ENABLE_MASK;
+	else
+		reg &= ~DCP_CLK_ENABLE_MASK;
+
+	io_write32(ccm_base + CCM_CCGR0, reg);
+}
+
+/*
+ * Use spinlock to secure clock register wrtiting
+ * Enable clock if it's not done
+ * Disable clock only if no channel is still using it
+ *
+ * @enable   Enable the clock if true, disable if false
+ */
+static void secure_dcp_clk_enable(bool enable)
+{
+	clock_except = cpu_spin_lock_xsave(&clock_spinlock);
+	if (enable) {
+		clock_refcount++;
+		if (clock_refcount == 1)
+			dcp_clk_enable(true);
+	} else {
+		clock_refcount--;
+		if (!clock_refcount)
+			dcp_clk_enable(false);
+	}
+	cpu_spin_unlock_xrestore(&clock_spinlock, clock_except);
+}
+
+/*
+ * Lock a channel using a mutex
+ *
+ * @chan   DCP channel to lock
+ */
+static TEE_Result dcp_lock_known_channel(enum dcp_channel chan)
+{
+	bool lock = false;
+	unsigned int timeout = 0;
+
+	for (timeout = 0; timeout < DCP_MAX_TIMEOUT; timeout++) {
+		lock = mutex_trylock(&lock_channel[chan].m);
```

Comment:
Why use a mutex if you not going to block on it?

Reply:
We use ```trylock``` because ```dcp_lock_known_channel()``` is used in ```dcp_lock_any_ready_channel()``` where we try to find an available channel.